core/src/credential.rs

```diff
@@ -0,0 +1,145 @@
+use std::{error::Error, time::{Instant, SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VCContext = [String];
+
+pub(crate) const JSON_LD_CONTEXT:  VCContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VCContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct IssuedCredential <'a> {
+    verifiable_credential:VerifiableCredential,
+    proof: *mut CredentialProof,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<&str, &str>,
+    context: VCContext,
+}
+
+trait VC {
+    fn getSubject(&self, key: &str) -> Result<String, Error> {
+        match self.credential_subject.get(key) {
+            Some(&value) => Ok(value.to_string()),
+            None => Err(Error::new(ErrorKind::Other, "Key not found")),
+        } 
+    }
+    fn getContext(&self) -> VCContext {
+        self.context.clone()
+    }
+}
+
+/// Default VC trait implementation for CredentialManager.
+/// Callers may re-implement this trait in their code by: impl VC for CredentialManager {/...code here.../}
+impl VC for CredentialManager {}
+
+impl <'a> CredentialManager <'a> {
+    /// returns CredentialManager cred_subject_template with default configuration accessible for overlapping via VC trait
+    pub fn new() -> CredentialManager {
+        ///default credential subjects for different credential types 
+        let cred_subject_template: HashMap<&str, &str> = HashMap::from([
+                (CRED_TYPE_PERMANENT_RESIDENT_CARD, r###"{
+                    "@id": "https://w3id.org/citizenship#PermanentResidentCard",
+                    "@context": {
+                      "@version": 1.1,
+                      "@protected": true,
+                      "id": "@id",
+                      "type": "@type",
+                      "description": "http://schema.org/description",
+                      "name": "http://schema.org/name",
+                      "identifier": "http://schema.org/identifier",
+                      "image": {"@id": "http://schema.org/image", "@type": "@id"}
+                    }
+                  }"###),
+                (CRED_TYPE_BANK_CARD, r###"{
```

Comment:
These attributes look like they belong to permanent resident card https://w3c-ccg.github.io/citizenship-vocab/#:~:text=3.-,The%20Citizenship%20Vocabulary,-This%20vocabulary%20assumes

Note the bankcard VC we have defined so far as a placeholder (Subject to change, but the w3c citizenship/permanent resident card above is not) is defined here https://www.notion.so/Verifiable-Credentials-Adapter-MVP-Design-56e1ea273c054fb39cc27ea1c8fcd744#8628511fbb99474aaa3617ee9eb00846